### PR TITLE
chore(ci): Fix new issues in cd.yml

### PR DESCRIPTION
**Pull request by jamilbk**
Fixes some issues encountered after the merge of #4049 

- Fix performance tests to only run using base_ref and head_ref to avoid dependence on `main`
- Fixes some typos
- Prevents a catch-22 condition where breaking compatibility meant we wouldn't be able to deploy production

### CODE REVIEW

**Comment by jamilbk** [comment]
Merging this because compare-results won't work without a successful CD run on `main`.

File: .github/workflows/ci.yml
Comment (thomaseizinger) [inline .github/workflows/ci.yml]: Would be nice to tidy this up and only use a regular matrix based on the name. Not blocking this PR of course.

File: .github/workflows/ci.yml
Comment (thomaseizinger) [inline .github/workflows/ci.yml:205]: So if this workflow is never run on `main`, how do we get results to compare?

**Review by thomaseizinger — COMMENTED** [review]
I don't think I understand how this fixes missing performance numbers for `main`.

File: .github/workflows/ci.yml
Comment (jamilbk) [reply .github/workflows/ci.yml:205]: We run performance tests on the `base_ref` which is more correct for cases where we've stacked the PR.

Comment (jamilbk) [reply .github/workflows/ci.yml:205]: Same trick I use to test DB migrations in CI:

- Checkout base_ref
- Run steps
- Check head_ref
- Re-run steps

It's a little bit of duplication but worth it IMO to avoid linking workflow runs together across branches which makes CI brittle.

File: .github/workflows/ci.yml
Comment (jamilbk) [reply .github/workflows/ci.yml]: Fixed

File: .github/workflows/ci.yml
Comment (thomaseizinger) [reply .github/workflows/ci.yml:205]: So we always run them twice per PR? Once for the base ref and once for the current one? Would it be worth to extract that into a workflow then that we call with two different refs? I think that would make things easier to understand.

Feel free to do in a follow-up if you want to get this merged.

Comment (thomaseizinger) [reply .github/workflows/ci.yml:205]: I think those two could also run in parallel which would be nice.

**Review by thomaseizinger — APPROVED** [review]
I understand now, I think the design is neat!

File: .github/workflows/ci.yml
Comment (jamilbk) [reply .github/workflows/ci.yml:205]: Ah yes, good call they should just be a matrix job

Comment (jamilbk) [reply .github/workflows/ci.yml:205]: Updated, this will merge when it passes.

**Pull request merged**
